pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/impl/FixedByteSingleColumnMultiValueReaderWriter.java

```diff
@@ -79,46 +107,52 @@
+  public FixedByteSingleColumnMultiValueReaderWriter(int maxNumberOfMultiValuesPerRow, int avgMultiValueCount, int rowCountPerChunk,
+      int columnSizeInBytes) {
+    int initialCapacity = Math.max(maxNumberOfMultiValuesPerRow, rowCountPerChunk * avgMultiValueCount);
     int incrementalCapacity =
         Math.max(maxNumberOfMultiValuesPerRow, (int) (initialCapacity * 1.0f * INCREMENT_PERCENTAGE / 100));
-    init(rows, columnSizeInBytes, maxNumberOfMultiValuesPerRow, initialCapacity, incrementalCapacity);
-  }
-
-  private void init(int rows, int columnSizeInBytes, int maxNumberOfMultiValuesPerRow, int initialCapacity,
-      int incrementalCapacity) throws IOException {
     this.columnSizeInBytes = columnSizeInBytes;
     this.maxNumberOfMultiValuesPerRow = maxNumberOfMultiValuesPerRow;
-    headerSize = rows * SIZE_OF_INT * NUM_COLS_IN_HEADER;
+    headerSize = rowCountPerChunk * SIZE_OF_INT * NUM_COLS_IN_HEADER;
+    this.rowCountPerChunk = rowCountPerChunk;
+    addHeaderBuffers();
+    //at least create space for million entries, which for INT translates into 4mb buffer
+    this.incrementalCapacity = incrementalCapacity;
+    addDataBuffers(initialCapacity);
+    //init(rowCountPerChunk, columnSizeInBytes, maxNumberOfMultiValuesPerRow, initialCapacity, incrementalCapacity);
+  }
+
+  private void addHeaderBuffers() {
     headerBuffer = PinotDataBuffer.allocateDirect(headerSize);
     // We know that these bufffers will not be copied directly into a file (or mapped from a file).
     // So, we can use native byte order here.
     headerBuffer.order(ByteOrder.nativeOrder());
     //dataBufferId, startIndex, length
-    headerWriter =
-        new FixedByteSingleValueMultiColWriter(headerBuffer, rows, 3,
+    curHeaderWriter =
+        new FixedByteSingleValueMultiColWriter(headerBuffer, rowCountPerChunk, 3,
             new int[] { SIZE_OF_INT, SIZE_OF_INT, SIZE_OF_INT });
-    headerReader =
-        new FixedByteSingleValueMultiColReader(headerBuffer, rows, new int[] { SIZE_OF_INT, SIZE_OF_INT, SIZE_OF_INT });
-    //at least create space for million entries, which for INT translates into 4mb buffer
-    this.incrementalCapacity = incrementalCapacity;
-    addCapacity(initialCapacity);
+    FixedByteSingleValueMultiColReader curHeaderReader =
+        new FixedByteSingleValueMultiColReader(headerBuffer, rowCountPerChunk, new int[] { SIZE_OF_INT, SIZE_OF_INT, SIZE_OF_INT });
+    headerBuffers.add(headerBuffer);
+    headerWriters.add(curHeaderWriter);
+    headerReaders.add(curHeaderReader);
   }
 
   /**
    * This method automatically computes the space needed based on the columnSizeInBytes
    * @param rowCapacity Additional capacity to be added in terms of number of rows
    * @throws RuntimeException
    */
-  private void addCapacity(int rowCapacity) throws RuntimeException {
+  private void addDataBuffers(int rowCapacity) throws RuntimeException {
```

Comment:
[saying aloud; no action needed] using FixedByteSingleValueMultiColWriter and writing one value at a time seems suboptimal for writing array of values. For nativeByte order, we should be able to use PinotDataBuffer and readFrom method for bulk write but layered interfaces are hurting us....